### ISSUE TITLE
m/[#]/x treated inconsistently

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9038]
Migrated from [rt.perl.org#45667](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=45667) (status was 'resolved')

Searchable as RT45667$

**Comment by p5pRT** [comment]
### From chris@chrisdolan.net

#### Created by cdolan@cpan.org



These should output the same value\, but instead the first one emits  
"f" and the second one undef\.

&nbsp;       my $qr = qr/\(f\)/;
&nbsp;       print "\#f" =~ m/\[\#\]\(f\)/x;
&nbsp;       print "\#f" =~ m/\[\#\]$qr/x;

This is more revealing&#8203;:

&nbsp;       print qr/\[\#\]$qr/x;

yields&#8203;:

&nbsp;       \(?x\-ism&#8203;:\[\#\]$qr
&nbsp;       \)

So\, it appears that \[\#\] is disabling interpolation for variables that  
follow\, but then still treating the content as regexp\.  Clearly\, it's  
not \*just\* treating it as a comment\, or the remaining qr/\[/ would be  
a syntax error\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=low

Site configuration information for perl v5.8.6:

Configured by root at Thu Nov  2 10:35:22 PST 2006.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
   Platform:
     osname=darwin, osvers=8.0, archname=darwin-thread-multi-2level
     uname='darwin b07.apple.com 8.0 darwin kernel version 8.3.0: mon  
oct 3 20:04:04 pdt 2005; root:xnu-792.6.22.obj~2release_ppc power  
macintosh powerpc '
     config_args='-ds -e -Dprefix=/usr -Dccflags=-g  -pipe  - 
Dldflags=-Dman3ext=3pm -Duseithreads -Duseshrplib'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define  
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-g -pipe -fno-common -DPERL_DARWIN -no-cpp- 
precomp -fno-strict-aliasing -I/usr/local/include',
     optimize='-Os',
     cppflags='-no-cpp-precomp -g -pipe -fno-common -DPERL_DARWIN -no- 
cpp-precomp -fno-strict-aliasing -I/usr/local/include'
     ccversion='', gccversion='3.3 20030304 (Apple Computer, Inc.  
build 1819)', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
     ivtype='long', ivsize=4, nvtype='double', nvsize=8,  
Off_t='off_t', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags ='-L/usr/ 
local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-ldbm -ldl -lm -lc
     perllibs=-ldl -lm -lc
     libc=/usr/lib/libc.dylib, so=dylib, useshrplib=true,  
libperl=libperl.dylib
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-bundle -undefined dynamic_lookup -L/ 
usr/local/lib'

Locally applied patches:
     23953 - fix for File::Path::rmtree CAN-2004-0452 security issue
     33990 - fix for setuid perl security issues
     SPRINTF0 - fixes for sprintf formatting issues - CVE-2005-3962


@INC for perl v5.8.6:
     /Users/chris/perl/lib/perl5/site_perl/5.8.6/darwin-thread- 
multi-2level
     /Users/chris/perl/lib/perl5/site_perl/5.8.6
     /Users/chris/perl/lib/perl5/site_perl
     /Users/chris/perl/lib/perl5/darwin-thread-multi-2level
     /Users/chris/perl/lib/perl5
     /Users/chris/perl/lib/perl5-core/5.8.6/darwin-thread-multi-2level
     /Users/chris/perl/lib/perl5-core/5.8.6
     /Users/chris/perl/lib/perl5-core
     /Users/chris/perl/lib/darwin-thread-multi-2level
     /Users/chris/perl/lib
     /Users/chris/perl/System/Library/Perl/5.8.6/darwin-thread- 
multi-2level
     /Users/chris/perl/System/Library/Perl/5.8.6
     /Users/chris/perl/System/Library/Perl
     /Users/chris/perl/System/Library/Perl/5.8.6/darwin-thread- 
multi-2level
     /Users/chris/perl/System/Library/Perl/5.8.6
     /sw/lib/perl5/5.8.6/darwin-thread-multi-2level
     /sw/lib/perl5/5.8.6
     /sw/lib/perl5/darwin-thread-multi-2level
     /sw/lib/perl5
     /sw/lib/perl5/darwin
     /System/Library/Perl/5.8.6/darwin-thread-multi-2level
     /System/Library/Perl/5.8.6
     /Library/Perl/5.8.6/darwin-thread-multi-2level
     /Library/Perl/5.8.6
     /Library/Perl
     /Network/Library/Perl/5.8.6/darwin-thread-multi-2level
     /Network/Library/Perl/5.8.6
     /Network/Library/Perl
     /System/Library/Perl/Extras/5.8.6/darwin-thread-multi-2level
     /System/Library/Perl/Extras/5.8.6
     /Library/Perl/5.8.1
     .


Environment for perl v5.8.6:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/chris
     LANG (unset)
     LANGUAGE (unset)
     LC_ALL=C
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/Users/chris/perl/bin:/Users/chris/perl/usr/bin:/Users/ 
chris/perl/script:/sw/bin:/sw/sbin:.:/Users/chris/bin:/usr/X11R6/bin:/ 
bin:/sbin:/usr/bin:/usr/sbin:/Developer/Tools
     PERL5LIB=/Users/chris/perl/lib/perl5/site_perl:/Users/chris/perl/ 
lib/perl5:/Users/chris/perl/lib/perl5-core:/Users/chris/perl/lib:/ 
Users/chris/perl/System/Library/Perl:/Users/chris/perl/System/Library/ 
Perl/5.8.6:/sw/lib/perl5:/sw/lib/perl5/darwin
     PERL_BADLANG (unset)
     PERL_CRITIC_CACHE=1
     SHELL=/bin/tcsh
	

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Sep 22\, 2007 at 08&#8203;:05&#8203;:59AM \-0700\, Chris Dolan wrote&#8203;:
> These should output the same value\, but instead the first one emits  
> "f" and the second one undef\.
> 
>        my $qr = qr/\(f\)/;
>        print "\#f" =~ m/\[\#\]\(f\)/x;
>        print "\#f" =~ m/\[\#\]$qr/x;
> 
> This is more revealing&#8203;:
> 
>        print qr/\[\#\]$qr/x;
> 
> yields&#8203;:
> 
>        \(?x\-ism&#8203;:\[\#\]$qr
>        \)
> 
> So\, it appears that \[\#\] is disabling interpolation for variables that  
> follow\, but then still treating the content as regexp\.  Clearly\, it's  
> not \*just\* treating it as a comment\, or the remaining qr/\[/ would be  
> a syntax error\.

Hmm\, looks like the tokenizer isn't smart enough to know to ignore \#'s
when within a character class\. S\_scan\_const\(\) needs to be made even more
scary than it already is\. Probably a post 5\.10\.0 job\.


\-\- 
"You may not work around any technical limitations in the software"
&nbsp;    \-\- Windows Vista license

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @demerphq

On 9/23/07\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> On Sat\, Sep 22\, 2007 at 08&#8203;:05&#8203;:59AM \-0700\, Chris Dolan wrote&#8203;:
> > These should output the same value\, but instead the first one emits
> > "f" and the second one undef\.
> >
> >        my $qr = qr/\(f\)/;
> >        print "\#f" =~ m/\[\#\]\(f\)/x;
> >        print "\#f" =~ m/\[\#\]$qr/x;
> >
> > This is more revealing&#8203;:
> >
> >        print qr/\[\#\]$qr/x;
> >
> > yields&#8203;:
> >
> >        \(?x\-ism&#8203;:\[\#\]$qr
> >        \)
> >
> > So\, it appears that \[\#\] is disabling interpolation for variables that
> > follow\, but then still treating the content as regexp\.  Clearly\, it's
> > not \*just\* treating it as a comment\, or the remaining qr/\[/ would be
> > a syntax error\.
>
> Hmm\, looks like the tokenizer isn't smart enough to know to ignore \#'s
> when within a character class\. S\_scan\_const\(\) needs to be made even more
> scary than it already is\. Probably a post 5\.10\.0 job\.

It is not clear to me that this is a bug actually\. The documentation
states that \# must be escaped in an /x pattern afair\.

Yves
\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Sep 24\, 2007 at 10&#8203;:05&#8203;:13AM \+0200\, demerphq wrote&#8203;:
> On 9/23/07\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > On Sat\, Sep 22\, 2007 at 08&#8203;:05&#8203;:59AM \-0700\, Chris Dolan wrote&#8203;:
> > > These should output the same value\, but instead the first one emits
> > > "f" and the second one undef\.
> > >
> > >        my $qr = qr/\(f\)/;
> > >        print "\#f" =~ m/\[\#\]\(f\)/x;
> > >        print "\#f" =~ m/\[\#\]$qr/x;
> > >
> > > This is more revealing&#8203;:
> > >
> > >        print qr/\[\#\]$qr/x;
> > >
> > > yields&#8203;:
> > >
> > >        \(?x\-ism&#8203;:\[\#\]$qr
> > >        \)
> > >
> > > So\, it appears that \[\#\] is disabling interpolation for variables that
> > > follow\, but then still treating the content as regexp\.  Clearly\, it's
> > > not \*just\* treating it as a comment\, or the remaining qr/\[/ would be
> > > a syntax error\.
> >
> > Hmm\, looks like the tokenizer isn't smart enough to know to ignore \#'s
> > when within a character class\. S\_scan\_const\(\) needs to be made even more
> > scary than it already is\. Probably a post 5\.10\.0 job\.
> 
> It is not clear to me that this is a bug actually\. The documentation
> states that \# must be escaped in an /x pattern afair\.

But it also says that \# isn't special when in a character class&#8203;:

&nbsp;    $ \./perl \-Dr \-e'/\[\#\]a
&nbsp;    b/x'
&nbsp;    Compiling REx "\[\#\]a%nb"
&nbsp;    rarest char \# at 0
&nbsp;    Final program&#8203;:
&nbsp;       1&#8203;: EXACT \<\#ab> \(5\)
&nbsp;       5&#8203;: END \(0\)

&nbsp;    $ perl \-le'print "ok" if "\#" =~ /\[\#\]/x'
&nbsp;    ok
&nbsp;    $



\-\- 
Britain\, Britain\, Britain\! Discovered by Sir Henry Britain in
sixteen\-oh\-ten\. Sold to Germany a year later for a pfennig and the promise
of a kiss\. Destroyed in eighteen thirty\-forty two\, and rebuilt a week
later by a man\. This we know\. Hello\. But what of the people of Britain?
Who they? What do? And why?   \-\- "Little Britain"

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Sep 24\, 2007 at 01&#8203;:52&#8203;:17PM \+0100\, Dave Mitchell wrote&#8203;:
> On Mon\, Sep 24\, 2007 at 10&#8203;:05&#8203;:13AM \+0200\, demerphq wrote&#8203;:
> > On 9/23/07\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > > On Sat\, Sep 22\, 2007 at 08&#8203;:05&#8203;:59AM \-0700\, Chris Dolan wrote&#8203;:
> > > > These should output the same value\, but instead the first one emits
> > > > "f" and the second one undef\.
> > > >
> > > >        my $qr = qr/\(f\)/;
> > > >        print "\#f" =~ m/\[\#\]\(f\)/x;
> > > >        print "\#f" =~ m/\[\#\]$qr/x;
> > > >
> > > > This is more revealing&#8203;:
> > > >
> > > >        print qr/\[\#\]$qr/x;
> > > >
> > > > yields&#8203;:
> > > >
> > > >        \(?x\-ism&#8203;:\[\#\]$qr
> > > >        \)
> > > >
> > > > So\, it appears that \[\#\] is disabling interpolation for variables that
> > > > follow\, but then still treating the content as regexp\.  Clearly\, it's
> > > > not \*just\* treating it as a comment\, or the remaining qr/\[/ would be
> > > > a syntax error\.
> > >
> > > Hmm\, looks like the tokenizer isn't smart enough to know to ignore \#'s
> > > when within a character class\. S\_scan\_const\(\) needs to be made even more
> > > scary than it already is\. Probably a post 5\.10\.0 job\.
> > 
> > It is not clear to me that this is a bug actually\. The documentation
> > states that \# must be escaped in an /x pattern afair\.
> 
> But it also says that \# isn't special when in a character class&#8203;:
> 
>     $ \./perl \-Dr \-e'/\[\#\]a
>     b/x'
>     Compiling REx "\[\#\]a%nb"
>     rarest char \# at 0
>     Final program&#8203;:
>        1&#8203;: EXACT \<\#ab> \(5\)
>        5&#8203;: END \(0\)
> 
>     $ perl \-le'print "ok" if "\#" =~ /\[\#\]/x'
>     ok
>     $

Do we have an agreed TODO test? \(or tests\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Sep 28 08&#8203;:07&#8203;:29 2007\, nicholas wrote&#8203;:
> On Mon\, Sep 24\, 2007 at 01&#8203;:52&#8203;:17PM \+0100\, Dave Mitchell wrote&#8203;:
> > On Mon\, Sep 24\, 2007 at 10&#8203;:05&#8203;:13AM \+0200\, demerphq wrote&#8203;:
> > > On 9/23/07\, Dave Mitchell \<davem@&#8203;iabyn\.com> wrote&#8203;:
> > > > On Sat\, Sep 22\, 2007 at 08&#8203;:05&#8203;:59AM \-0700\, Chris Dolan wrote&#8203;:
> > > > > These should output the same value\, but instead the first one
> emits
> > > > > "f" and the second one undef\.
> > > > >
> > > > >        my $qr = qr/\(f\)/;
> > > > >        print "\#f" =~ m/\[\#\]\(f\)/x;
> > > > >        print "\#f" =~ m/\[\#\]$qr/x;
> > > > >
> > > > > This is more revealing&#8203;:
> > > > >
> > > > >        print qr/\[\#\]$qr/x;
> > > > >
> > > > > yields&#8203;:
> > > > >
> > > > >        \(?x\-ism&#8203;:\[\#\]$qr
> > > > >        \)
> > > > >
> > > > > So\, it appears that \[\#\] is disabling interpolation for
> variables that
> > > > > follow\, but then still treating the content as regexp\.
> Clearly\, it's
> > > > > not \*just\* treating it as a comment\, or the remaining qr/\[/
> would be
> > > > > a syntax error\.
> > > >
> > > > Hmm\, looks like the tokenizer isn't smart enough to know to
> ignore \#'s
> > > > when within a character class\. S\_scan\_const\(\) needs to be made
> even more
> > > > scary than it already is\. Probably a post 5\.10\.0 job\.
> > >
> > > It is not clear to me that this is a bug actually\. The
> documentation
> > > states that \# must be escaped in an /x pattern afair\.
> >
> > But it also says that \# isn't special when in a character class&#8203;:
> >
> >     $ \./perl \-Dr \-e'/\[\#\]a
> >     b/x'
> >     Compiling REx "\[\#\]a%nb"
> >     rarest char \# at 0
> >     Final program&#8203;:
> >        1&#8203;: EXACT \<\#ab> \(5\)
> >        5&#8203;: END \(0\)
> >
> >     $ perl \-le'print "ok" if "\#" =~ /\[\#\]/x'
> >     ok
> >     $
> 
> Do we have an agreed TODO test? \(or tests\)

Agreed?  Probably not\.  I’ve been thinking about this issue lately\,
specifically in conjunction with \(but not limited to\) \(?\[\.\.\.\]\) regexp sets\.

/x mode allows variables to be commented out&#8203;:

$ \./perl \-Ilib \-le '$x="\\n"; print "a" =~ / a \# $x blah blah /x'
1

But not always&#8203;:

$ \./perl \-Ilib \-Xle '$\# = "\#"; $x="\\n"; print / \[a\] $\# $x blah blah /x
|| nope'
nope

Regexp sets are supposed to imply /x\, but variables are not commented out&#8203;:

$ \./perl \-Xe '$x="\\n"; /\(?\[ \[a\] \# $x blah blah' \-e'\]\)/'
Unexpected character in regex; marked by \<\-\- HERE in m/\(?\[ \[a\] \# 
 b \<\-\- HERE lah blah
\]\)/ at \-e line 1\.

Remove the variable\, and it is fine&#8203;:

$  \./perl \-Xe '$x="\\n"; /\(?\[ \[a\] \# blah blah' \-e '\]\)/'
$

Or add /x; that is also fine&#8203;:

$ \./perl \-Xe '$x="\\n"; /\(?\[ \[a\] \# $x blah blah' \-e'\]\)/x'
$

So this is a bigger can of worms than anyone thought\.  Perhaps the real
fix is for the regexp engine to call back to the parser when it sees
what it thinks is a variable\.  Dave Mitchell is not going to like what I
am suggesting\. :\-\)

Also\, I wonder whether this is at all fixable in the end\.  I suspect
however hard we try we will just end up fixing a few cases and breaking
a few others\, because the rules are ultimately contradictory\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Jun 22\, 2013 at 07&#8203;:03&#8203;:17PM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> So this is a bigger can of worms than anyone thought\.  Perhaps the real
> fix is for the regexp engine to call back to the parser when it sees
> what it thinks is a variable\.  Dave Mitchell is not going to like what I
> am suggesting\. :\-\)

Um\, lets not do that\!

\-\- 
No man treats a motor car as foolishly as he treats another human being\.
When the car will not go\, he does not attribute its annoying behaviour to
sin\, he does not say\, You are a wicked motorcar\, and I shall not give you
any more petrol until you go\. He attempts to find out what is wrong and
set it right\.
&nbsp;    \-\- Bertrand Russell\,
&nbsp;       Has Religion Made Useful Contributions to Civilization?

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jun 24 06&#8203;:50&#8203;:43 2013\, davem wrote&#8203;:
> On Sat\, Jun 22\, 2013 at 07&#8203;:03&#8203;:17PM \-0700\, Father Chrysostomos via RT
wrote&#8203;:
> > So this is a bigger can of worms than anyone thought\.  Perhaps the real
> > fix is for the regexp engine to call back to the parser when it sees
> > what it thinks is a variable\.  Dave Mitchell is not going to like what I
> > am suggesting\. :\-\)
> 
> Um\, lets not do that\!

In that case we will have to accept that only literal \# can prevent
interpolation\.  The remainder of this bug \(the original report\) was
fixed by c30fc27b4df65a\, and I have just added tests in 61965f1\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
